Formal2024/Workshops/workshop4.lean

```diff
@@ -0,0 +1,259 @@
+/-
+Copyright (c) 2024 TheLeanTeam. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: The Lean Team
+-/
+import Library
+open Set
+
+/-!
+  # Sets and Functions
+  References: [MIL] Mathematics in Lean, [Tut] Tutorials project.
+  Most of the demonstration section comes from [MIL].
+-/
+
+section
+
+-- Let `α` be a generic type
+variable {α : Type*}
+
+-- We denote by `Set α` the type of sets of elements of `α`
+variable (s t u : Set α)
+#check s
+
+end
+-- This may be confusing, as we were intuitively thinking of types as sets
+
+/-
+  # What actually is a type?
+-/
+section
+-- Type is the primitive notion in the Calculus of Inductive Constructions, upon which Lean is based
+-- Every term has a type
+
+#check 1
+#check 1+1=2 -- We discussed how `Prop` is the type of propositions
+
+#check ℕ -- The type of `ℕ` is the type of types, called `Type`
+#check Prop -- Same for `Prop`
+#check Type -- Lean has a countably infinite hierarchy of type of types (of types of types...)
+#check Type 1
+--#check Type 33 -- Well, almost infinite
+-- unless you manually increase the `maxUniverseOffset`
+set_option maxUniverseOffset 33
+#check Type 33
+
+#check Prop → ℕ -- We can construct types of functions between types
+#check Type 1 → Type 2
+
+-- There are two inference rules that tell us that functions work like we would expect
+
+variable {α β : Type*} -- With the asterisk we are leaving the level of the type generic
+#check α
+#check β
+variable (f : α → β)
+variable (a : α)
+
+#check f a -- We can apply functions, getting the expected outcome
+
+def myAnd (P : Prop) : Prop := P ∧ P
+
+#check fun (P: Prop) ↦ myAnd P -- We can perform abstractions to create functions
+
+
+-- Apart from `∀` (which we discussed is equivalent to functions) everything else is constructed
+-- starting from here (mostly using Inductive Types, as we will see).
+end
+
+
+section
+/-
+  # How is a type different from a set?
+-/
+
+def setNat : Set ℕ := univ -- This is the set of all natural numbers
+
+-- Sets come with a predicate: `∈` (`\in` or `\mem`). Its negation is `∉` (`\notin`)
+
+#check ( 1 ∈ setNat ) -- This is in `Prop`
+#check ( 1 : ℕ )     -- This is not
+
+-- `( 1 : ℕ )` is a statement in type theory, but not an element in `Prop`, our model of
+-- propositional logic inside type theory.
+
+-- Another big difference: if `a : α` and `a : β`, then `α` and `β` are definitionally equal
```

Comment:
It's from that paper you shared on Discord "The type theory of Lean". Theorem 4.1 (Unique typing)

Reply:
Hmm... might make sense in theory, but definitely not in practice. Idk, I can't make much sense of it.

Reply:
```lean
example {α β : Type*} (a : α) (a : β) : α = β := ?
```

Reply:
This makes me think of the following as a good exercise (for this or next week):

```lean
example : ∃ α β : Type, α ≠ β
```

Reply:
> Hmm... might make sense in theory, but definitely not in practice. Idk, I can't make much sense of it.

From the way it looks it is a meta-theorem, not sure if you can prove it in Lean. I guess the point I wanted to make is that for sets we have subsets, but there is no analogue for types. 
I should add a more grounded example like how (1 : \N) is different from (1 : \R)

Reply:
Yes! Say that exactly. In fact, write something along the lines of:

```lean
-- if you do
-- `example : (1 : ℕ) = (1 : ℤ)`
-- you will see an error about mismatched types and applications and all
```

Reply:
If you do want to keep the original sentence, then I suggest you add that it is a "meta-theorem" (so cannot be proved in Lean), and also add a reference

Reply:
> I guess the point I wanted to make is that for sets we have subsets, but there is no analogue for types. I should add a more grounded example like how (1 : \N) is different from (1 : \R)

But make sure when you say this, you say 1 as a natural number cannot equal to 1 as a real number (because if you write the opposite in Lean, it just automatically gets coerced into the reals), and we don't want to confuse them with coercions just yet haha

Reply:
> Yes! Say that exactly. In fact, write something along the lines of:
> 
> ```lean
> -- if you do
> -- `example : (1 : ℕ) = (1 : ℤ)`
> -- you will see an error about mismatched types and applications and all
> ```
It seems like Lean has no problem with this, you can even solve it with norm_num ahaha

Reply:
hahah. alright, this gives an error:
```lean
example : (1 : ℕ) = ((1 : ℝ) : ℤ)
```

1 as a natural number equals 1 as an integer (which is 1 as a real number) gives an error